lib/similarity_utils.py

```diff
@@ -118,9 +134,131 @@ def print_top_similiar(main_actor: str, names: List[str], values: List[float]) -
+    dla każdego wiersza liczy wartość similarity pomiędzy pierwotnie istniejącym w wierszu aktorem, a aktorem wskazanym
+    w argumencie main_actor_values"""
+    columns = ["tconst", "titleType", "genres", "category"]
+    main_columns = []
+    for column in columns:
+        main_columns.append(column + "_main")
+        data = insert_main_actor_column_values(data, column, main_actor_values[column])
+    similarities = list(data.apply(similarity_pandas, axis=1))
+    data.drop(columns = main_columns, axis=1)
+    return list(data["nconst"]), similarities
+
+
+def normalized_manhattan_distance(list1: List[float], list2: List[float]) -> float:
+    """metoda przyjmuje jako argumenty listy o wartościach liczbowych oraz zwraca liczbę która jest znormalizowaną (długością list)
+    odległość Manhattan pomiedzy podanymi dwoma listami"""
+    if not list1 or not list2:
+        return 0
+    else:
+        length = len(list1)
+        assert length == len(list2), "listy dla których liczona jest odległość są różnych długości"
+        distance = sum(abs(value1 - value2) for value1, value2 in zip(list1, list2))
+        return distance / length
+
+
+def similarity_new(actor1: List[Any], actor2: List[Any], reduced_dataset: bool = False) -> float:
+    """metoda licząca similarity pomiędzy dwoma aktorami;
+    jej argumentami są dwie listy, a wartością wyjściową wartość z przedziału [-1, 1];
+    parametr 'reduced_dataset' domyślnie ustawiony jest na False, jednak w przypadku gdy zbiór 
+    danych na którym liczymy similarity ograniczamy do listy poniższych 12 kolumn:
+    
+    ["nconst", "tconst", "category", "primaryName", "knownForTitles", "no_nominations_oscars_norm", "no_nominations_globes_norm",
+    "no_nominations_emmy_norm", "no_films_norm", "average_films_rating_norm", "genres_code", "types_code"]
+    
+    to należy zmienić wartość parametru 'reduced_dataset' na True;
+    metoda jest przygotowana pod dane ze zbiorów treningowego, testowego i walidacyjnego"""
+    weights = [0.2, 0.03, 0.13, 0.03, 0.03, 0.03, 0.12, 0.13, 0.15, 0.15]
+    if reduced_dataset:
+        values = [
+            iou(actor1[1], actor2[1]),          # similarity ze względu na ilość wspólnych filmów
+            1 if actor1[2] == actor2[2] else 0, # similarity ze względu na tę samą płeć
+            iou(actor1[4], actor2[4]),          # similarity ze względu na wspólne produkcje z których są znani
+            1 - abs(actor1[5] - actor2[5]),    # similarity ze względu na ilość nominacji do oscarów
+            1 - abs(actor1[6] - actor2[6]),    # similarity ze względu na ilość nominacji do globów
+            1 - abs(actor1[7] - actor2[7]),    # similarity ze względu na ilość nominacji do emmy
+            1 - abs(actor1[8] - actor2[8]),   # similarity ze względu na ilość zagranych filmów
+            1 - abs(actor1[9] - actor2[9]),   # similarity ze względu na średnią ocenę filmów w których grali aktorzy
```

Comment:
czy tu nie ma możliwosci, że różnica będzie na tyle duża, że similaritywyskoczy poza [-1,1]?

Reply:
Jeżeli wszystkie wartości w kolumnach z końcówką `_norm` są znormalizowane do przedziału [0,1] to nie ma takiej szansy.

Reply:
Wszystkie wartości z values są w zakresie [0,1], a suma wag wynosi 1. Stąd suma ich odpowiednich iloczynów zawsze będzie w przedziale [0,1]. Po zadziałaniu mapowaniem f(x) = 2*x -1 otrzymamy similarity z przedziału [-1,1].